rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/http_body_0_4_x.rs

```diff
@@ -114,288 +145,45 @@ where
-                body,
-                MinimumThroughputBodyOptions::builder()
-                    .minimum_throughput(minimum_throughput)
-                    .build(),
-            ))
-        });
-
-        (body.collect(), time_source, async_sleep)
-    }
+impl<B> Body for ThroughputReadingBody<B>
+where
+    B: Body<Data = bytes::Bytes, Error = BoxError>,
+{
+    type Data = bytes::Bytes;
+    type Error = BoxError;
 
-    async fn expect_error(minimum_throughput: Throughput) {
-        let (res, ..) =
-            eight_byte_per_second_stream_with_minimum_throughput_timeout(minimum_throughput);
-        let expected_err = Error::ThroughputBelowMinimum {
-            expected: minimum_throughput,
-            actual: Throughput::new(8, Duration::from_secs(1)),
-        };
-        match res.await {
-            Ok(_) => {
-                panic!(
-                    "response succeeded instead of returning the expected error '{expected_err}'"
-                )
-            }
-            Err(actual_err) => {
-                assert_eq!(
-                    expected_err.to_string(),
-                    // We need to source this so that we don't get the streaming error it's wrapped in.
-                    actual_err.source().unwrap().to_string()
-                );
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        // this code is called quite frequently in production—one every millisecond or so when downloading
+        // a stream. However, SystemTime::now is on the order of nanoseconds
+        let now = self.time_source.now();
+        // Attempt to read the data from the inner body, then update the
+        // throughput logs.
+        let this = self.as_mut().project();
+        match this.inner.poll_data(cx) {
+            Poll::Ready(Some(Ok(bytes))) => {
+                tracing::trace!("received data: {}", bytes.len());
+                this.throughput
+                    .push_bytes_transferred(now, bytes.len() as u64);
+                Poll::Ready(Some(Ok(bytes)))
             }
-        }
-    }
-
-    #[tokio::test]
-    async fn test_throughput_timeout_less_than() {
-        let minimum_throughput = Throughput::new_bytes_per_second(9);
-        expect_error(minimum_throughput).await;
-    }
-
-    async fn expect_success(minimum_throughput: Throughput) {
-        let (res, time_source, async_sleep) =
-            eight_byte_per_second_stream_with_minimum_throughput_timeout(minimum_throughput);
-        match res.await {
-            Ok(res) => {
-                assert_eq!(255.0, time_source.seconds_since_unix_epoch());
-                assert_eq!(Duration::from_secs(255), async_sleep.total_duration());
-                assert_eq!(*EXPECTED_BYTES, res.to_vec());
+            Poll::Pending => {
+                tracing::trace!("received poll pending");
+                this.throughput.push_pending(now);
```

Comment:
seems like we could slightly refactor the API to be something like:

`this.throughput.record_poll(poll_result)`

```
fn record_poll(&mut self, p: Poll<Bytes>) -> Poll<Bytes>
```